biz.aQute.bndlib/src/aQute/bnd/junit/JUnitFramework.java

```diff
@@ -0,0 +1,331 @@
+import aQute.bnd.build.Run;
+import aQute.bnd.build.Workspace;
+import aQute.bnd.header.Attrs;
+import aQute.bnd.header.Parameters;
+import aQute.bnd.osgi.Analyzer;
+import aQute.bnd.osgi.Builder;
+import aQute.bnd.osgi.Descriptors.PackageRef;
+import aQute.bnd.osgi.Jar;
+import aQute.bnd.osgi.JarResource;
+import aQute.bnd.osgi.Resource;
+import aQute.bnd.osgi.URLResource;
+import aQute.bnd.service.Strategy;
+import aQute.bnd.version.VersionRange;
+import aQute.lib.exceptions.Exceptions;
+import aQute.lib.io.IO;
+
+/**
+ * This class provides an OSGi framework that is configured with the current bnd
+ * workspace. A project directory is used to find the workspace. This makes all
+ * repositories in the workspace available to the framework. To be able to test
+ * JUnit code against/in this framework it is necessary that all packages on the
+ * buildpath and testpath are actually exported in the framework. This class
+ * will ensure that. Once the framework is up and running it will be possible to
+ * add bundles to it. There are a number of ways that this can be achieved:
+ * <ul>
+ * <li>Build a bundle – A bnd {@link Builder} is provided to create a bundle and
+ * install it. This makes it possible to add classes from the src or test
+ * directories or resources. See {@link #bundle()}.
+ * <li>Add a bundle using a bnd spec – Using the bnd specs (e.g.
+ * 'org.apache.felix.configadmin;version=3'). See {@link #addBundle(String)} et.
+ * al.
+ * <li>Add a bndrun file – A file bndrun file can be added. All properties in
+ * this file that can be applied after a framework is started will be applied.
+ * See {@link #addBundles(File)} et. al.
+ * </ul>
+ * Convenience methods are added to get services, see {@link #getService(Class)}
+ * et. al. Notice that this framework starts in the same process as that the
+ * JUnit code runs. This is normally a separately started VM.
+ */
+public class JUnitFramework implements AutoCloseable {
+	ExecutorService								executor	= Executors.newCachedThreadPool();
+	public final List<ServiceTracker< ? , ? >>	trackers	= new ArrayList<>();
+	public final Jar							bin_test;
+	public final Framework						framework;
+	public final BundleContext					context;
+	public final File							projectDir;
+	public Workspace							workspace;
+	public Project								project;
+
+	/**
+	 * Start a framework assuming the current working directory is the project
+	 * directory.
+	 */
+	public JUnitFramework() {
+		this(IO.work);
+	}
+
+	/**
+	 * Start a framework while providing a project directory.
+	 * 
+	 * @param projectDir
+	 */
+	public JUnitFramework(File projectDir) {
+		this.projectDir = projectDir;
+
+		try {
+			Project p = getProject();
+			File bin_test = p.getTestOutput();
+			this.bin_test = new Jar(bin_test);
+
+			String extra = getExtra();
+
+			Map<String,String> props = new HashMap<>();
+			props.put(Constants.FRAMEWORK_SYSTEMPACKAGES_EXTRA, extra);
+
+			File storage = IO.getFile("generated/fw");
```

Comment:
This is a fixed folder name in the current dir. Seems like a bad idea. Since you have the project, you should make a folder in the target folder.